src/utils/aLinkCreator/doc/alcc.1
.TH ALCC 1 "March 2005" "aLinkCreator" "aMule utilities"
.SH NAME
aLinkCreator \- the aMule ED2k link creator
.SH SYNOPSIS
.B alcc
.RB [ \-h ]
.RB [ \-v ]
.RB [ \-p ]
.I input files...
.SH DESCRIPTION
\fBalcc\fR is a command-line utility to create an ED2k link to any file on your computer.
.TP
\fB\-h\fR, \fB\-\-help\fR
show usage help message
.TP
\fB\-v\fR, \fB\-\-verbose\fR
be verbose
.TP
\fB\-p\fR, \fB\-\-parthashes\fR
add part-hashes to ed2k link
.SH REPORTING BUGS
Please report bugs either on our forum (\fIhttp://forum.amule.org/\fR), or in our bugtracker (\fIhttp://bugs.amule.org/\fR).
Please do not report bugs in e-mail, neither to our mailing list nor directly to any team member.
.SH COPYRIGHT
aMule and all of its related utilities are distributed under the GNU General Public License.
.SH SEE ALSO
\fBalc\fR(1), \fBamule\fR(1)


ChangeLog
Version 2.3.4 - The "can we have two releases in the same year?" version.
----------
2021-??-??


--------------------------------------------------------------------------------

Version 2.3.3 - The "delayed too many times, but still not too late" version.
----------
2021-02-07

	alesnav:
		* Permit amuleweb in an iframe

	Dmitry Smirnov:
		* Added mimetype-hanlder to amule.desktop
		  originally in debian

	gaffatape:
		* Workaround for bug in wxWidgets causing aMule to crash on
		  closing the last search tab

	GonoszTopi:
		* Fix restoring toolbar orientation on 'Cancel'
		* Features ported from eMule:
			- Find best keyword for Kad search (0.49c)
		* Various fixes to gdlib detection
		* Support for compiling aMule with wxGTK under Cygwin
		* Various small fixes to the MacOSX packager script
		* Workaround for wxWidgets bug #17740 (Wrong path used to
		  search for translations), fixes GitHub issue #74 (.mo files
		  for translation are not loaded if amule is built to install
		  to a prefix other than /usr (e.g. /usr/local).)
		* Portability fixes
		* Handle the deprecation of std::auto_ptr in C++11 by using
		  std::unique_ptr instead.
		* Fix linking with readline if configure uses caching
		* Remove 'web search' search type as there are no search
		  engines to support
		* Readline related fixes and improvements in amulecmd:
			- Set the proper application name (amulecmd) for
			  readline. This allows conditional parsing of the
			  readline configuration files (like ~/.inputrc). See
			  the GNU Readline Library documentation (section 1.3.2)
			  for more details.
			- Set up command completion on pressing the TAB key.
			  Since command names are case insensitive, completion
			  uses the default capitalization (shown by the 'Help'
			  command). Completion also works for the
			  'Help <command name>' construct.
			- Pressing <Enter> on an empty line won't give 'Syntax
			  error!' anymore, the empty line is silently ignored.
		* Flex and bison generated files will no longer be regenerated
		  occasionally on the user's system.
		* Proxy authorization fixes
		* Explicitly disable support for autoconf >= 2.70
		* Fix compilation using C++17

	jEsuSdA:
		* Added papirus theme

	KeyofBlueS:
		* Fix compilation with newer bfd

	kinote:
		* Fix integer conversions in amulecmd

	persmule:
		* Fix sharing directories with non-ASCII names with aMule daemon
		* Fix potential invalid detection for empty wxCharBuffer

	Phoenix:
		* Fixed libupnp API breakage.

	poplite:
		* Fixed issue #99 by properly initializing wxSockets in the
		  remote gui

	Sandro Tosi:
		* remove garbage from gdlib
		  originally in debian

	sc0w:
		* Fixed compilation with upnp-1.8
		* C++11 fixes
		* Only handle mouse-up events if the mouse-down event happened
		  on the same tab
		* Fix issue #114: Switching search result tabs doesn't work
		* Fix a number of build warnings
		* Fix a few memory leaks
		* Various small code improvements
		* Add option to show notifications when finished downloading
		* Fixed issue #54: the minimize button doesn't work propertly
		  when "minimize to tray icon" is activated on preferences
		* Fixed issue #59: start minimized doesn't work
		* Fixed issue #225: amule shows wrong 100% when files aren't
		  downloaded
		* Show speeds in MB/s instead of kB/s when appropriate

	Sebastiano Pistore:
		* Unified translation file headers to make it easier to identify
		  contributors

	Thibaut Lassalle:
		* Add README.md

	Tom Li:
		* OScopeCtrl.cpp: use Refresh() instead of ProcessEvent() to
		  refresh the UI, fix issue #146.

	Tommy Jerry Mairo:
		* Fixed issue #119: API mismatch with Crypto++ 6.0.0

	Vollstrecker:
		* Initial support for detecting pkg-config based gdlib versions
		* Compilation fixes for VS2015
		* Fixed compilation with wx-3.1
		* Added cmake-support (Linux and Windows tested)
		* Dropped support for cryptopp versions before 5.5

	xinhuang:
		* Fix crash when shared files changed on Shared files tab

--------------------------------------------------------------------------------

Version 2.3.2 - The "don't close tabs" version.
----------
2016-09-16

	circulosmeos:
		* amuleweb cookies read behind a "lowercase" proxy (http/2 compliant)

	Dan64:
		* Read AICH root hashes from binary eMuleCollection files

	gnazio:
		* Fixed configure of libpng when sh != bash

	GonoszTopi:
		* Allow any kind of file links in collections
		* Make UPnP log go to the appropriate category (#1681)
		* Use the C++ compiler to check for features we're about to use
		  in C++ source files (#1572)
		* Adapt configure to handle subversion repository clones in
		  mercurial and git
		* Fix bfd detection for systems missing -liberty
		* Add support for 'canceled.met' to the fileview tool
		* Implemented the missing 'show shared' command in amulecmd
		* Fix check for a valid locale config entry being done *after*
		  the locale has been set
		* Fix Crypto++ detection routine happily using build library
		  for host when cross-compiling
		* Fix bug #1711: Kad Info tab should display my own user ID
		* Fix wrong sized images in the priscilla skin
		* Fix bug #1423: Add the possibility to not create sparse files
		* Fix bug #1743: crash when importing part files
		* Review assertions in the eD2k and Kademlia code and remove
		  those that could be triggered by a malformed incoming packet
		* Fix a possible uninitialized variable access and a definite
		  memory leak in the UPnP code

	iz0bbz:
		* Fix WinSock library version detection with MinGW-w64
		* Fix compilation with MinGW-w64

	kam:
		* WebServer serves .json, .manifest and .appacache files

	Mr Hyde:
		* Fix "Disable computer's timed standby mode" for MacOS 10.9 Mavericks
		* Compilation fix on Linux Slackware 64bit and MacOS 10.9.4
		* Fix configure on Mac

	sc0w:
		* Add ability to use middle-click to close search tabs

	Stu Redman:
		* Display hashing progress in progress bars

src/utils/aLinkCreator/src/alcc.cpp
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/// Name:         Main wxBase App
///
/// Purpose:      aMule ed2k link creator
///
/// Author:       ThePolish <thepolish@vipmail.ru>
///
/// Copyright (c) 2004-2011 ThePolish ( thepolish@vipmail.ru )
///
/// This program is free software; you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation; either version 2 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program; if not, write to the
/// Free Software Foundation, Inc.,
/// 51 Franklin St, Fifth Floor, Boston, MA  02110-1301, USA
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////



#include "config.h"		// Needed for PACKAGE

#include <wx/log.h>

#include "alcc.h"
#include "ed2khash.h"

// Application implementation
IMPLEMENT_APP (alcc)

/// Running Alcc
int alcc::OnRun ()
{
  // Used to tell alcc to use aMule catalog
  m_locale.Init();
  m_locale.AddCatalog(PACKAGE);

  wxLog::DontCreateOnDemand();
  wxLogStderr * stderrLog = new wxLogStderr;
  wxLogStderr * stdoutLog = new wxLogStderr(stdout);
  delete wxLog::SetActiveTarget(stderrLog); // Log on Stderr
  wxLog::SetTimestamp("");   // Disable timestamp on messages

  Ed2kHash hash;
  size_t i;
  for (i=0;i<(m_filesToHash.GetCount());++i)
    {
      if (wxFileExists(m_filesToHash[i]))
        {
          if (m_flagVerbose)
            {
              wxLogMessage(_("Processing file number %u: %s"),i+1,m_filesToHash[i].c_str());

              if (m_flagPartHashes)
                {
                  wxLogMessage(_("You have asked for part hashes (Only used for files > 9.5 MB)"));
                }
            }

          if (hash.SetED2KHashFromFile(m_filesToHash[i], NULL))
            {
				// Print the link to stdout
				wxLog::SetActiveTarget(stdoutLog);
                wxLogMessage("%s", hash.GetED2KLink(m_flagPartHashes).c_str());
				// Everything else goes to stderr
				wxLog::SetActiveTarget(stderrLog);
            }
        }
      else
        {
            if (m_flagVerbose)
                {
                    wxLogMessage(_("%s ---> Non existent file !\n"),m_filesToHash[i].c_str());
                }
        }
    }
  delete stdoutLog;
  return 0;
}

// On exit
int
alcc::OnExit()
{
  delete wxLog::SetActiveTarget(NULL);
  return 0;
}

/// Parse command line
void alcc::OnInitCmdLine(wxCmdLineParser& cmdline)
{
	cmdline.AddSwitch("h", "help", "show this help message", wxCMD_LINE_OPTION_HELP);
	cmdline.AddSwitch("v", "verbose", "be verbose");
	cmdline.AddSwitch("p", "parthashes", "add part-hashes to ed2k link");
	cmdline.AddParam("input files", wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_MULTIPLE);
}

/// Command line preprocessing
bool alcc::OnCmdLineParsed(wxCmdLineParser& cmdline)
{

  wxFileName filename;
  size_t i;

  m_flagVerbose = cmdline.Found("v");
  m_flagPartHashes = cmdline.Found("p");

  m_filesToHash.Clear();
  for (i = 0; i < cmdline.GetParamCount(); ++i)
    {
      filename.Assign(cmdline.GetParam(i));
      m_filesToHash.Add(filename.GetFullPath());
    }
  m_filesToHash.Shrink();

  return true;
}
// File_checked_for_headers
